fix e2e onboarding e2e test when verticalization is enabled

roo2 opened the pull request:
#### Proposed Changes

Currently, when v13n is enabled, the test `specs/onboarding/signup__free` will fail. This test is not run as part of CI but needs to be run as part of the Pre-Release tests in TeamCity.

fix for goals screen https://github.com/Automattic/wp-calypso/pull/64829/files#diff-5fd720308095682e2cd4100ab6078deebb8c86ea241524478028c3bbe754bff7R77
slack discussion p1655851613524289-slack-CRWCHQGUB

Note: currently it appears that the `signup__paid` flow hasn't been working for a while? I'm working on fixing this as well and updating it to support the new v13n flow

#### Testing Instructions
run the e2e test locally 
```
yarn install
cd test/e2e
export E2E_SECRETS_KEY='Calypso E2E Config decode key from the Automattic secret store>'
yarn decrypt-secrets
 
yarn workspace @automattic/calypso-e2e build

# run the tests against different environments
CALYPSO_BASE_URL=https://horizon.wordpress.com yarn jest specs/onboarding
CALYPSO_BASE_URL=http://calypso.localhost:3000 yarn jest specs/onboarding
CALYPSO_BASE_URL=https://wordpress.com yarn jest specs/onboarding

roo2 committed:
fix specs/onboarding/signup__free test for when verticals is enabled

roo2 commented on packages/calypso-e2e/src/lib/flows/start-site-flow.ts:
```diff
@@ -43,20 +45,38 @@ export class StartSiteFlow {
+		const defaultInputlocator = this.page.locator( selectors.blogNameInput );
+		// try waiting for element to become enabled
+		// we've had issues with this before https://github.com/Automattic/wp-calypso/issues/64271
+		waitForElementEnabled( this.page, selectors.blogNameInput );
```
Since the input no longer has "Site Title" pre-filled, we can't use this selector anymore. This is similar to the old approach before https://github.com/Automattic/wp-calypso/issues/64271 where we waited for `enabled` state. I wasn't able to reproduce the issues myself, I think we'd better just see if they occur on CI. Open to other ideas for how to handle this 🤔

worldomonation replied:
The issue with the site name and tagline resetting is fixed in https://github.com/Automattic/wp-calypso/pull/64590.

worldomonation replied:
Since the issue in #64590 has been fixed, you shouldn't need to use `waitForElementEnabled` - in fact I would consider `waitForElementEnabled` to be deprecated so would not encourage its use. I just haven't gotten around to dealing with it yet.

Alternatively can I suggest using locators?

```
const inputLocator = page.locator( selector );
await inputLocator.fill( name );
```

roo2 replied:
Ahh great, I think I didn't understand the other comment and thought it was an issue purely related to the e2e tests.

So, I think I should add the `not(:disabled)` selectors back in as it was before https://github.com/Automattic/wp-calypso/pull/64458/files right?

worldomonation replied:
@roo2 

> So, I think I should add the not(:disabled) selectors back in as it was before

Right, the `input#siteTitle:not(:disabled)` selector should work right out of the box. The changes made in #64458 failed to make the script stable which was (as noted earlier) caused by a real bug.

worldomonation commented on test/e2e/specs/onboarding/signup__free.ts:
```diff
@@ -81,7 +81,20 @@ describe( DataHelper.createSuiteTitle( 'Signup: WordPress.com Free' ), function
+		} );
+
+		it( 'Select a vertical', async function () {
+			const isVerticalsStep = await page.isVisible( '.site-vertical' );
```
I wrote a long message in https://github.com/Automattic/wp-calypso/pull/64829/files#r904406971 before I realized the changes had been merged, but generally speaking, the `isVisible` approach is _very_ flaky and should not be relied upon.

In the linked comment I made a suggestion which uses `locators`, which will greatly improve the reliability of the script compared to using `isVisible`. I'd appreciate it if you could try it out.

worldomonation replied:
Note, if my request is unclear, let's drop the change request for now - in the interest of helping iterate quicker. 

I can work on moving the checks into the page objects, since that's one of my responsibilities.

worldomonation reviewed (changes requested):
Overall, a reasonable change but I'd like to have at least these two addressed prior to merge:
- use of isVisible
- use of `waitForElementEnabled`

Ideally, I would like [this](https://github.com/Automattic/wp-calypso/pull/64829/files#r904408665) to also be addressed, but it would require changes to the `StartSiteFlow` page object which is a larger task and so I am comfortable leaving that as another task for either myself or someone else to handle.

roo2 commented:
thanks @worldomonation I will look at replacing isVisible as you suggested. and move the `waitForElementEnabled` check into the selector

I'm happy to make the change you suggested in https://github.com/Automattic/wp-calypso/pull/64829/files#r904408665 here as well.

Thanks for the Feedback!

worldomonation commented on packages/calypso-e2e/src/lib/flows/start-site-flow.ts:
```diff
@@ -43,20 +45,38 @@ export class StartSiteFlow {
+	 *
+	 * @param {string} vertical Name of the vertical to select
+	 */
+	async enterVertical( vertical: string ): Promise< void > {
+		await this.page.waitForLoadState( 'networkidle' );
+
+		const input = this.page.locator( selectors.verticalInput );
+		await input.fill( vertical );
+
+		const readBack = await input.inputValue();
+		if ( readBack !== vertical ) {
+			throw new Error( `Failed to set vertical: expected ${ vertical }, got ${ readBack }` );
+		}
+	}
```
**praise**: this is good implementation, very similar to what I would do 👍

roo2 replied:
I think I just copied the pattern from a function you probably wrote in the same file 😂. But yes, I can see it is a nice pattern to check whether the input was filled correctly.

roo2 committed:
PR feedback

roo2 commented:
@worldomonation this is ready for another look. I moved the check for which page we are on into the flow object and used locators rather than `isVisible` I also removed the `isEnabled` check and moved that back to the selector.

Note that we need to support several versions of this flow for now
Goals & Verticalization enabled => horizon.wordpress.com
Goals only => wpcalypso.wordpress.com
Neither => wordpress.com

That is the reason for the complex logic that needs to know the current step

worldomonation reviewed (approved):
Triggered Pre-Release Tests against the feature branch and it appears to pass. 

As with any changes to the Pre-Release Tests, I will be keeping a close eye on the stability of changes to make sure we don't introduce flakiness.

taipeicoder reviewed (approved):
Tested and it LGTM 👍
